crates/core/src/host/module_host.rs

```diff
@@ -668,45 +671,98 @@ impl ModuleHost {
-            // an empty transaction to ensure we always have those events
-            // paired in the commitlog.
-            //
-            // This is necessary to be able to disconnect clients after a server
-            // crash.
-            ReducerCallError::NoSuchReducer => {
-                let db = &self.inner.dbic().relational_db;
-                db.with_auto_commit(
-                    &ExecutionContext::reducer(
-                        db.address(),
-                        ReducerContext {
-                            name: reducer_name.to_owned(),
-                            caller_identity,
-                            caller_address,
-                            timestamp: Timestamp::now(),
-                            arg_bsatn: Bytes::new(),
-                        },
-                    ),
-                    |_| anyhow::Ok(()),
-                )
-                .ok();
-
-                Ok(())
-            }
-            e => Err(e),
-        })
+        let db = &self.inner.dbic().relational_db;
+        let ctx = &ExecutionContext::reducer(
+            db.address(),
+            ReducerContext {
+                name: reducer_name.to_owned(),
+                caller_identity,
+                caller_address,
+                timestamp: Timestamp::now(),
+                arg_bsatn: Bytes::new(),
+            },
+        );
+
+        let result = self
+            .call_reducer_inner(
+                caller_identity,
+                Some(caller_address),
+                None,
+                None,
+                None,
+                reducer_name,
+                ReducerArgs::Nullary,
+            )
+            .await
+            .map(drop)
+            .or_else(|e| match e {
+                // If the module doesn't define connected or disconnected, commit
+                // a transaction to update `st_clients` and to ensure we always have those events
+                // paired in the commitlog.
+                //
+                // This is necessary to be able to disconnect clients after a server
+                // crash.
+                ReducerCallError::NoSuchReducer => db
+                    .with_auto_commit(ctx, |mut_tx| {
+                        if connected {
+                            self.update_st_clients(mut_tx, caller_identity, caller_address, connected)
+                        } else {
+                            Ok(())
+                        }
+                    })
+                    .map_err(|err| {
+                        InvalidReducerArguments {
+                            err: err.into(),
+                            reducer: reducer_name.into(),
+                        }
+                        .into()
+                    }),
+                e => Err(e),
+            });
+
+        // Deleting client from `st_clients`does not depend upon result of disconnect reducer hence done in a separate tx.
```

Comment:
This will create two `__identity_disconnected__` entries in the commitlog in the case disconnect is defined and succeeds.

That's fine for the purpose of the auto-disconnect mechanism, but we shouldn't be afraid of over-documenting it. E.g. offline analysis like @lcodes is doing might find it surprising.

Maybe just add commentary to the enclosing method explaining what is going on.

Reply:
The behavior is correct, though, without support for nested transactions.

Reply:
Agreed, makes sense.